retrieval/discovery/kubernetes.go

```diff
@@ -0,0 +1,582 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.package discovery
+
+package discovery
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"sync"
+
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+	"github.com/prometheus/log"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/retrieval/discovery/kubernetes"
+	"github.com/prometheus/prometheus/util/httputil"
+)
+
+const (
+	kubernetesSourcePrefix        = "kubernetes"
```

Comment:
The prefixing is usually fine for variables and constant, as long as it's done. But sooner or later internal helper functions can become messy and we probably don't want to prefix those.

I think having sub-packages in `retrieval/discovery/<sd_mechanism>` with the implementations and the constructors in `retrieval/discovery` would be nice.